irohad/validation/impl/chain_validator_impl.cpp

```diff
@@ -30,41 +20,64 @@ namespace iroha {
         : supermajority_checker_(supermajority_checker),
           log_(logger::log("ChainValidator")) {}
 
-    bool ChainValidatorImpl::validateBlock(
-        std::shared_ptr<shared_model::interface::Block> block,
-        ametsuchi::MutableStorage &storage) const {
-      log_->info("validate block: height {}, hash {}",
-                 block->height(),
-                 block->hash().hex());
-      auto check_block =
-          [this](const auto &block, auto &queries, const auto &top_hash) {
-            auto peers = queries.getLedgerPeers();
-            if (not peers) {
-              return false;
-            }
-            return block.prevHash() == top_hash
-                and supermajority_checker_->hasSupermajority(block.signatures(),
-                                                             peers.value());
-          };
-
-      // check inside of temporary storage
-      return storage.check(*block, check_block);
-    }
-
     bool ChainValidatorImpl::validateChain(
         rxcpp::observable<std::shared_ptr<shared_model::interface::Block>>
             blocks,
         ametsuchi::MutableStorage &storage) const {
       log_->info("validate chain...");
-      return blocks
-          .all([this, &storage](auto block) {
-            log_->info("Validating block: height {}, hash {}",
-                       block->height(),
-                       block->hash().hex());
-            return this->validateBlock(block, storage);
-          })
-          .as_blocking()
-          .first();
+
+      return storage.apply(
+          blocks,
+          [this](const auto &block, auto &queries, const auto &top_hash) {
+            return this->applyBlock(block, queries, top_hash);
+          });
+    }
+
+    bool ChainValidatorImpl::applyBlock(
```

Comment:
I think, new name `applyBlock` is misleading as well - the function does not apply the block, it just checks if it's applicable. Maybe, leave it `validateBlock` or, for instance, `canApplyBlock`?